### DISCUSSION TITLE
Dealing with XML that is sometimes an array, and sometimes not

### DISCUSSION THREAD

**Discussion by goose-ws** [mikefarah/yq#2106] [Q&A] [answered]
I'm using `yq` to parse some data from API lookups in Plex. The specific string I'm using it:

    yq -p xml '.MediaContainer.Directory | select ( ."+@index" == "2024" ) | ."+@ratingKey"'

The trouble is, sometimes my `.Directory` is an array, and sometimes it is not, depending on how many seasons of an item are available. Is there some one-line way I can add `.[]` to the end of `.Directory` if it's an array, and omit it if it's not?

Thanks in advance :)

**Comment by mikefarah** [comment, marked as answer]
Yep, take a look at the examples here: https://mikefarah.gitbook.io/yq/usage/xml#parse-xml-force-as-an-array. Let me know if you still need help :)

**Reply by goose-ws** [reply]
Thanks for the super quick reply! I've been playing around with the example for `force as an array` that you linked, but I've been unsuccessful so far in trying to manipulate it to do what I want.

For a non-array item, I'm able to extract the `.MediaContainer.Directory.+@ratingKey" value that I want via:

    yq -p xml ".MediaContainer.Directory |= ([] + .) | select ( .MediaContainer.Directory[0].\"+@index\" == \"2024\" ) | .MediaContainer.Directory[0].\"+@ratingKey\""

But for an array item, I haven't yet figured out how to string it together with the `|= ([] + .)` modifier to make it pull the `+@ratingKey` for the specific `+@index` year that I want.

I'm probably also not using `yq` as efficiently as it can be for this purpose, either. I'd appreciate any suggestions you have on how to better format my commands.

**Edit:**

Further banging my head against the wall has found that:

    yq -p xml ".MediaContainer.Directory | ([] + .) | .[] | select ( .\"+@index\" == \"2024\" ) .\"+@ratingKey\""

Does what I want it to do for both array and non-array items. Although I'd still appreciate any feedback on more efficient use of `yq`.

**Reply by mikefarah** [reply]
Without having a sample file, it's a little hard to tell. This looks pretty good efficiency wise - I'm a little suspicious of this bit: ` ([] + .) | .[] ` as it seems like you're putting into an array, only to explode it again. You might be able to get away with just `.[]`. That said, this doesn't make a big difference though efficiency wise.
